src/utilstrencodings.cpp

```diff
@@ -19,15 +19,19 @@ static const std::string SAFE_CHARS[] =
     CHARS_ALPHA_NUM + " .,;-_/:?@()", // SAFE_CHARS_DEFAULT
     CHARS_ALPHA_NUM + " .,;-_?@", // SAFE_CHARS_UA_COMMENT
     CHARS_ALPHA_NUM + ".-_", // SAFE_CHARS_FILENAME
+    CHARS_ALPHA_NUM + " .,;-_/:?@()!\"#$%&'*+<=>[\\]^`{|}~"  // SAFE_CHARS_PRINTABLE
 };
 
-std::string SanitizeString(const std::string& str, int rule)
+std::string SanitizeString(const std::string& str, int rule, bool escape)
 {
     std::string strResult;
     for (std::string::size_type i = 0; i < str.size(); i++)
     {
-        if (SAFE_CHARS[rule].find(str[i]) != std::string::npos)
+        if (SAFE_CHARS[rule].find(str[i]) != std::string::npos) {
             strResult.push_back(str[i]);
+        } else if (escape) {
+            strResult += strprintf("%%%02x", str[i]);
```

Comment:
You will also want to remove the `%` character from `SAFE_CHARS_PRINTABLE` so it will be escaped itself, or alternately add a `&& (!escape || str[i] != '%')` clause to the find check above.